rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/CharacteristicLongWriteOperation.java

```diff
@@ -209,4 +218,30 @@ public Boolean call(Object emission) {
             }
         };
     }
+
+    private Action1<ByteAssociation<UUID>> resetRetryCounter() {
+        return new Action1<ByteAssociation<UUID>>() {
+            @Override
+            public void call(ByteAssociation<UUID> uuidByteAssociation) {
+                retryCounter = 0;
+            }
+        };
+    }
+
+    private Func2<Integer, Throwable, Boolean> retryStrategy(final ByteBuffer byteBuffer, final int batchSize) {
+        return new Func2<Integer, Throwable, Boolean>() {
+            @Override
+            public Boolean call(Integer integer, Throwable throwable) {
+                // Individual counter for each batch payload.
+                retryCounter++;
+                final Boolean retry = writeOperationRetryStrategy.call(retryCounter, throwable);
```

Comment:
The biggest question here is to:
1) Define what data one will need to decide on a retry
2) Smart way to extend that data without breaking the API in the future (DTO? @dariuszseweryn)

Reply:
Yup, probably we should go for `WriteOperationRetryStrategy` with something like:
```
class LongWriteFailure {
  final int batchNumber;
  // final int retryCount; –> this should not be needed as it is trivial for the user to manage
  final BleException cause;
}
```

Reply:
You're reading in my mind :)

Reply:
I came up with the following `retryWhen()` function:
```java
private static Func1<Observable<? extends Throwable>, Observable<?>> retryOperationStrategy(
            final WriteOperationRetryStrategy writeOperationRetryStrategy,
            final ByteBuffer byteBuffer,
            final int batchSize) {
        return new Func1<Observable<? extends Throwable>, Observable<?>>() {
            @Override
            public Observable<?> call(Observable<? extends Throwable> observable) {
                return observable
                        .flatMap(applyRetryStrategy())
                        .map(replaceByteBufferPositionForRetry());
            }

            @NonNull
            private Func1<Throwable, Observable<WriteOperationRetryStrategy.LongWriteFailure>> applyRetryStrategy() {
                return new Func1<Throwable, Observable<WriteOperationRetryStrategy.LongWriteFailure>>() {
                    @Override
                    public Observable<WriteOperationRetryStrategy.LongWriteFailure> call(Throwable cause) {
                        if (!(cause instanceof BleException)) {
                            return Observable.error(cause);
                        }

                        final int failedBatchNumber = (byteBuffer.position() - batchSize) / batchSize;
                        final WriteOperationRetryStrategy.LongWriteFailure longWriteFailure =
                                new WriteOperationRetryStrategy.LongWriteFailure(failedBatchNumber, (BleException) cause);

                        return writeOperationRetryStrategy.call(Observable.just(longWriteFailure));
                    }
                };
            }

            @NonNull
            private Func1<WriteOperationRetryStrategy.LongWriteFailure, Object> replaceByteBufferPositionForRetry() {
                return new Func1<WriteOperationRetryStrategy.LongWriteFailure, Object>() {
                    @Override
                    public Object call(WriteOperationRetryStrategy.LongWriteFailure longWriteFailure) {
                        final int newBufferPosition = longWriteFailure.getBatchNumber() * batchSize;
                        byteBuffer.position(newBufferPosition);
                        return longWriteFailure;
                    }
                };
            }
        };
    }
```

Reply:
If you could push another commit to this PR I could then comment on specific lines

Reply:
Sure, done!